File: packages/core/src/logger.ts
import { prettyPrint } from './pretty-print'

const LEVELS = {
  NOTSET: 0,
  DEBUG: 10,
  INFO: 20,
  WARNING: 30,
  ERROR: 40,
  CRITICAL: 50,
} as const

const levelMap: Record<string, number> = {
  debug: LEVELS.DEBUG,
  info: LEVELS.INFO,
  warn: LEVELS.WARNING,
  warning: LEVELS.WARNING,
  error: LEVELS.ERROR,
  critical: LEVELS.CRITICAL,
}

const getLogLevel = (): number => {
  const level = process.env.LOG_LEVEL ?? 'info'
  return levelMap[level] ?? LEVELS.INFO
}

const shouldLog = (messageLevel: number): boolean => {
  return messageLevel >= getLogLevel()
}

export type LogListener = (level: string, msg: string, args?: unknown) => void

export class Logger {
  /**
   * Why do we need two level of listeners?
   *
   * Core listeners pass along to children loggers.
   *
   * However, base listeners do not pass along to children loggers.
   * Those are specific to each logger in the hierarchy.
   */
  private readonly listeners: LogListener[] = []

  constructor(
    readonly isVerbose: boolean = false,
    private readonly meta: Record<string, unknown> = {},
    private readonly coreListeners: LogListener[] = [],
  ) {}

  child(meta: Record<string, unknown>): Logger {
    return new Logger(this.isVerbose, { ...this.meta, ...meta }, this.coreListeners)
  }

  private _log(level: string, msg: string, args?: any) {
    const time = Date.now()
    const meta = { ...this.meta, ...(args ?? {}) }
    prettyPrint({ level, time, msg, ...meta }, !this.isVerbose)

    this.coreListeners.forEach((listener) => listener(level, msg, meta))
    this.listeners.forEach((listener) => listener(level, msg, meta))
  }

  info(message: string, args?: unknown) {
    if (shouldLog(LEVELS.INFO)) {
      this._log('info', message, args)
    }
  }

  error(message: string, args?: unknown) {
    if (shouldLog(LEVELS.ERROR)) {
      this._log('error', message, args)
    }
  }

  debug(message: string, args?: unknown) {
    if (shouldLog(LEVELS.DEBUG)) {
      this._log('debug', message, args)
    }
  }

  warn(message: string, args?: unknown) {
    if (shouldLog(LEVELS.WARNING)) {
      this._log('warn', message, args)
    }
  }

  log(args: any) {
    const level = args.level ?? 'info'
    const messageLevel = levelMap[level] ?? LEVELS.INFO

    if (!shouldLog(messageLevel)) return

    this._log(level, args.msg, args)
  }

  addListener(listener: LogListener) {
    this.listeners.push(listener)
  }

  removeListener(listener: LogListener) {
    const index = this.listeners.indexOf(listener)
    if (index > -1) {
      this.listeners.splice(index, 1)
    }
  }
}

export const globalLogger = new Logger()

File: packages/core/src/python/motia_core/logger.py
import os
import time
from typing import Callable, Dict, Any, Optional, List
from .pretty_print import pretty_print

LOG_LEVEL = os.getenv("LOG_LEVEL", "info").lower()

_is_debug_enabled = LOG_LEVEL == "debug"
_is_info_enabled = LOG_LEVEL in ("info", "debug")
_is_warn_enabled = LOG_LEVEL in ("warn", "info", "debug", "trace")

LogListener = Callable[[str, str, Optional[Dict[str, Any]]], None]

class Logger:
    def __init__(
        self,
        is_verbose: bool = False,
        meta: Optional[Dict[str, Any]] = None,
        core_listeners: Optional[List[LogListener]] = None,
    ) -> None:
        self.is_verbose = is_verbose
        self._meta = dict(meta or {})
        self._core_listeners: List[LogListener] = list(core_listeners or [])
        self._listeners: List[LogListener] = []

    def child(self, meta: Dict[str, Any]) -> "Logger":
        merged = {**self._meta, **(meta or {})}
        return Logger(self.is_verbose, merged, self._core_listeners)

    def _log(self, level: str, msg: str, args: Optional[Dict[str, Any]] = None) -> None:
        now_ms = int(time.time() * 1000)
        meta = {**self._meta, **(args or {})}
        pretty_print({"level": level, "time": now_ms, "msg": msg, **meta}, exclude_details=not self.is_verbose)

        for listener in self._core_listeners:
            listener(level, msg, meta)
        for listener in self._listeners:
            listener(level, msg, meta)

    def info(self, message: str, args: Optional[Dict[str, Any]] = None) -> None:
        if _is_info_enabled:
            self._log("info", message, args)

    def error(self, message: str, args: Optional[Dict[str, Any]] = None) -> None:
        self._log("error", message, args)

    def debug(self, message: str, args: Optional[Dict[str, Any]] = None) -> None:
        if _is_debug_enabled:
            self._log("debug", message, args)

    def warn(self, message: str, args: Optional[Dict[str, Any]] = None) -> None:
        if _is_warn_enabled:
            self._log("warn", message, args)

    def log(self, args: Dict[str, Any]) -> None:
        msg = str(args.get("msg", ""))
        self._log("info", msg, args)

    def add_listener(self, listener: LogListener) -> None:
        self._listeners.append(listener)